src/utils/log_redact.py

```
@@ -0,0 +1,409 @@
+        # ``formatException`` and caches the result in ``record.exc_text``
+        # — at which point the raw exception message (e.g.
+        # ``RuntimeError("api_key=sk-xxx")``) is rendered verbatim into
+        # the log output. Pre-format and redact the traceback now so the
+        # formatter's "use cached exc_text if already set" branch picks
+        # up our scrubbed version instead. Critical: this must run even
+        # in the args branch (codex P1:
+        # ``logger.error("msg: %s", value, exc_info=True)``).
+        if record.exc_info:
+            try:
+                exc_text = "".join(traceback.format_exception(*record.exc_info))
+            except Exception:
```

Ruff (0.15.10)

[warning] Do not catch blind exception: `Exception`

(BLE001)